File: output/GPT4o-Mini/xCodeEval_4000/C++/C/e4411d4393b9c0c76c612cc437611459.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <math.h>
#include <stdbool.h>

typedef long long int int64;

void touch_pos(size_t r, size_t c, size_t d, const size_t *len, size_t **dist, size_t *q, size_t *front, size_t *rear) {
    size_t pos_c = (c < len[r]) ? c : len[r];
    if (d < dist[r][pos_c]) {
        dist[r][pos_c] = d;
        q[(*rear)++] = r;
        q[(*rear)++] = pos_c;
    }
}

void c253() {
    size_t nrow;
    scanf("%zu", &nrow);

    size_t *len = (size_t *)malloc(nrow * sizeof(size_t));
    for (size_t i = 0; i < nrow; i++) {
        scanf("%zu", &len[i]);
    }

    size_t first[2], last[2];
    scanf("%zu %zu", &first[0], &first[1]);
    scanf("%zu %zu", &last[0], &last[1]);

    first[0]--;
    first[1]--;
    last[0]--;
    last[1]--;

    if (first[0] == last[0] && first[1] == last[1]) {
        printf("0\n");
        free(len);
        return;
    }

    const size_t INF = SIZE_MAX;
    size_t **dist = (size_t **)malloc(nrow * sizeof(size_t *));
    for (size_t j = 0; j < nrow; j++) {
        dist[j] = (size_t *)malloc((len[j] + 1) * sizeof(size_t));
        for (size_t k = 0; k <= len[j]; k++) {
            dist[j][k] = INF;
        }
    }

    size_t *q = (size_t *)malloc(nrow * len[0] * 2 * sizeof(size_t));
    size_t front = 0, rear = 0;

    q[rear++] = first[0];
    q[rear++] = first[1];
    dist[first[0]][first[1]] = 0;

    while (front < rear) {
        size_t r = q[front++];
        size_t c = q[front++];
        size_t d = dist[r][c] + 1;

        if (r > 0 && last[0] == touch_pos(r - 1, c, d, len, dist, q, &front, &rear)) break;
        if (r + 1 < nrow && last[0] == touch_pos(r + 1, c, d, len, dist, q, &front, &rear)) break;
        if (c > 0 && last[0] == touch_pos(r, c - 1, d, len, dist, q, &front, &rear)) break;
        if (c < len[r] && last[0] == touch_pos(r, c + 1, d, len, dist, q, &front, &rear)) break;
    }

    printf("%zu\n", dist[last[0]][last[1]]);

    for (size_t j = 0; j < nrow; j++) {
        free(dist[j]);
    }
    free(dist);
    free(len);
    free(q);
}

int main() {
    c253();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/e4411d4393b9c0c76c612cc437611459.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func touchPos(r, c, d int, len []int, dist [][]int, q *[]int) (int, int) {
	pos := []int{r, int(math.Min(float64(c), float64(len[r])))}
	if d < dist[r][pos[1]] {
		dist[r][pos[1]] = d
		*q = append(*q, pos...)
	}
	return pos[0], pos[1]
}

func c253() {
	var n int
	fmt.Scan(&n)

	len := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&len[i])
	}

	var first, last []int
	fmt.Scan(&first)
	fmt.Scan(&last)

	first[0]--
	first[1]--
	last[0]--
	last[1]--

	if first[0] == last[0] && first[1] == last[1] {
		fmt.Println(0)
		return
	}

	INF := math.MaxInt32
	dist := make([][]int, n)
	for i := 0; i < n; i++ {
		dist[i] = make([]int, len[i]+1)
		for j := 0; j <= len[i]; j++ {
			dist[i][j] = INF
		}
	}

	q := make([]int, 0)
	firstPos := []int{first[0], int(math.Min(float64(first[1]), float64(len[first[0]])))}
	q = append(q, firstPos...)
	dist[first[0]][firstPos[1]] = 0

	for len(q) > 0 {
		r := q[0]
		c := q[1]
		d := dist[r][c] + 1
		q = q[2:]

		if r > 0 && last[0] == touchPos(r-1, c, d, len, dist, &q) {
			break
		}

		if r+1 < n && last[0] == touchPos(r+1, c, d, len, dist, &q) {
			break
		}

		if c > 0 && last[0] == touchPos(r, c-1, d, len, dist, &q) {
			break
		}

		if c < len[r] && last[0] == touchPos(r, c+1, d, len, dist, &q) {
			break
		}
	}

	fmt.Println(dist[last[0]][last[1]])
}

func main() {
	c253()
}
